sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -743,83 +875,79 @@ static void *_lwmac_thread(void *args)
     /* initialize low-level driver */
     dev->driver->init(dev);
 
-    /* Enable RX- and TX-started interrupts  */
-    netopt_enable_t enable = NETOPT_ENABLE;
     dev->driver->set(dev, NETOPT_RX_START_IRQ, &enable, sizeof(enable));
     dev->driver->set(dev, NETOPT_TX_START_IRQ, &enable, sizeof(enable));
     dev->driver->set(dev, NETOPT_TX_END_IRQ, &enable, sizeof(enable));
 
-    uint16_t src_len = 8;
-    dev->driver->set(dev, NETOPT_SRC_LEN, &src_len, sizeof(src_len));
-
     /* Get own address from netdev */
     gnrc_netdev->l2_addr_len = dev->driver->get(dev, NETOPT_ADDRESS_LONG,
                                                 &gnrc_netdev->l2_addr,
                                                 IEEE802154_LONG_ADDRESS_LEN);
     assert(gnrc_netdev->l2_addr_len > 0);
 
+    netif->mac.tx.bcast_seqnr = gnrc_netdev->l2_addr[0];
+#endif
+
+    /* RTT is used for scheduling wakeup */
+    rtt_init();
+
+    /* Store pid globally, so that IRQ can use it to send msg */
+    lwmac_pid = netif->pid;
+
+    /* Enable RX- and TX-started interrupts  */
+    netopt_enable_t enable = NETOPT_ENABLE;
+    netif->dev->driver->set(netif->dev, NETOPT_RX_START_IRQ, &enable, sizeof(enable));
+    netif->dev->driver->set(netif->dev, NETOPT_TX_START_IRQ, &enable, sizeof(enable));
+    netif->dev->driver->set(netif->dev, NETOPT_TX_END_IRQ, &enable, sizeof(enable));
+
+    uint16_t src_len = IEEE802154_LONG_ADDRESS_LEN;
+    netif->dev->driver->set(netif->dev, NETOPT_SRC_LEN, &src_len, sizeof(src_len));
+
     /* Initialize broadcast sequence number. This at least differs from board
      * to board */
-    gnrc_netdev->tx.bcast_seqnr = gnrc_netdev->l2_addr[0];
+    netif->mac.tx.bcast_seqnr = netif->l2addr[0];
 
     /* Reset all timeouts just to be sure */
-    gnrc_lwmac_reset_timeouts(gnrc_netdev);
+    gnrc_lwmac_reset_timeouts(netif);
 
     /* Start duty cycling */
-    lwmac_set_state(gnrc_netdev, GNRC_LWMAC_START);
+    lwmac_set_state(netif, GNRC_LWMAC_START);
 
 #if (GNRC_LWMAC_ENABLE_DUTYCYLE_RECORD == 1)
     /* Start duty cycle recording */
-    gnrc_netdev->lwmac.system_start_time_ticks = rtt_get_counter();
-    gnrc_netdev->lwmac.last_radio_on_time_ticks = gnrc_netdev->lwmac.system_start_time_ticks;
-    gnrc_netdev->lwmac.awake_duration_sum_ticks = 0;
-    gnrc_netdev->lwmac.lwmac_info |= GNRC_LWMAC_RADIO_IS_ON;
+    netif->mac.lwmac.system_start_time_ticks = rtt_get_counter();
+    netif->mac.lwmac.last_radio_on_time_ticks = netif->mac.lwmac.system_start_time_ticks;
+    netif->mac.lwmac.awake_duration_sum_ticks = 0;
+    netif->mac.lwmac.lwmac_info |= GNRC_LWMAC_RADIO_IS_ON;
```

Comment:
All of the `lwmac`-related code can go into `gnrc_netif2_ops_t::init()` so you don't need the thread anymore.

Reply:
Hi, Martine, what do you mean I don't need the thread anymore? Do you mean, (after I implemented `ops-parameters` with LwMAC features), there will be no LwMAC thread anymore (and there will be only the thread that created in `sys/net/gnrc/netif2/gnrc_netif2.c`), ?

Reply:
So, is the idea like this: LwMAC should be ported to the `sys/net/gnrc/netif2/gnrc_netif2.c` (the thread, maybe the MAC thread then?), with replacement of the `ops-parameters` (like `_init()`, `_send`...) of my own LwMAC implementation?

Reply:
Yes. That is one of the main features of `gnrc_netif2`, to simplify MAC porting and to avoid code duplication ;-).

Reply:
(you've done this this is just not as visible in the diff so everything alright here ;-))